File: src/client/acontext-ts/src/types/learning-space.ts
/**
 * Type definitions for learning space resources.
 */

import { z } from 'zod';

export const LearningSpaceSchema = z.object({
  id: z.string(),
  user_id: z.string().nullable().optional(),
  meta: z.record(z.string(), z.unknown()).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type LearningSpace = z.infer<typeof LearningSpaceSchema>;

export const LearningSpaceSkillSchema = z.object({
  id: z.string(),
  learning_space_id: z.string(),
  skill_id: z.string(),
  created_at: z.string(),
});

export type LearningSpaceSkill = z.infer<typeof LearningSpaceSkillSchema>;

/**
 * Valid status values for a learning space session.
 *
 * Lifecycle: pending → distilling → (skill_writing | queued | completed | failed)
 *
 * Sync: keep in sync with:
 *   - Python Core: src/server/core/acontext_core/schema/session/learning_space.py (SessionStatus)
 *   - Go API:      src/server/api/go/internal/modules/model/learning_space.go (SessionStatus* consts)
 *   - Python SDK:   src/client/acontext-py/src/acontext/types/learning_space.py (SessionStatus)
 */
export const SESSION_STATUSES = [
  'pending',
  'distilling',
  'queued',
  'skill_writing',
  'completed',
  'failed',
] as const;

export type SessionStatus = (typeof SESSION_STATUSES)[number];

/** Terminal statuses that indicate learning is complete. */
export const TERMINAL_SESSION_STATUSES: ReadonlySet<string> = new Set<SessionStatus>([
  'completed',
  'failed',
]);

export const LearningSpaceSessionSchema = z.object({
  id: z.string(),
  learning_space_id: z.string(),
  session_id: z.string(),
  status: z.enum(SESSION_STATUSES).or(z.string()),
  created_at: z.string(),
  updated_at: z.string(),
});

export type LearningSpaceSession = z.infer<typeof LearningSpaceSessionSchema>;

export const ListLearningSpacesOutputSchema = z.object({
  items: z.array(LearningSpaceSchema),
  next_cursor: z.string().nullable().optional(),
  has_more: z.boolean(),
});

export type ListLearningSpacesOutput = z.infer<typeof ListLearningSpacesOutputSchema>;

File: src/server/core/acontext_core/schema/session/learning_space.py
from enum import StrEnum


class SessionStatus(StrEnum):
    """Learning space session status enum.

    Tracks the lifecycle of a session being learned by a learning space:
    pending → distilling → (skill_writing | queued | completed | failed)
      - skill_writing → completed | failed
      - queued → distilling (re-enters via drain_skill_learn_pending)

    Sync: keep in sync with:
      - Go API:    src/server/api/go/internal/modules/model/learning_space.go (SessionStatus* consts)
      - Python SDK: src/client/acontext-py/src/acontext/types/learning_space.py (SessionStatus)
      - TS SDK:    src/client/acontext-ts/src/types/learning-space.ts (SESSION_STATUSES)
    """

    PENDING = "pending"
    DISTILLING = "distilling"
    QUEUED = "queued"
    SKILL_WRITING = "skill_writing"
    COMPLETED = "completed"
    FAILED = "failed"


TERMINAL_STATUSES = frozenset({SessionStatus.COMPLETED, SessionStatus.FAILED})
